app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -85,24 +114,32 @@ def remove_provider_from_session(provider)
       session.delete('omniauth_provider')
     end
 
-    # More info at:
-    # https://github.com/heartcombo/devise#omniauth
+    # We use the configured OmniAuth camelization to include the user-facing name of the provider in the flash message
+    def set_flash(type, key, kind: OmniAuth::Utils.camelize(omniauth_provider), reason: '')
+      flash[type] = t(key, kind:, reason:) if is_flashing_format?
+    end
 
-    # GET|POST /resource/auth/twitter
-    # def passthru
-    #   super
-    # end
+    def set_current_user
+      # Check if an existing user is already signed in (passed through the RelayState)
+      # and trying to add a new identity to their account. If so, we load the user information
+      # and set it as the current user. This is necessary to avoid creating a new user.
+      @current_user = User.find_by(id: OmniAuth::NonceStore.pop(params[:RelayState]))
+    end
 
-    # GET|POST /users/auth/twitter/callback
-    # def failure
-    #   super
-    # end
+    def set_user_identity
+      @user_identity = UserIdentity.includes(:user).find_or_initialize_by(omniauth_provider:, provider_uid:)
```

Comment:
At first, I was a bit skeptical about this change, since I feel we should align `set_user_identity` and `set_current_user`. While your suggested change works easily for `set_user_identity`, overwriting the `current_user` method comes at a greater risk (it is the default devise-defined method to access the signed-in user). After careful consideration, I thus went ahead and tried the following, which works as expected:

```rb
      @current_user ||= if params[:RelayState]
                          User.find_by(id: OmniAuth::NonceStore.pop(params[:RelayState])) || super
                        else
                          super
                        end
```

Would you suggest to use this approach?

Reply:
That's true. I will change that.